src/Analysis/Engine/Impl/DependencyResolution/PathResolverSnapshot.cs

```diff
@@ -91,48 +112,146 @@ public PathResolverSnapshot(PythonLanguageVersion pythonLanguageVersion)
+        private static bool TryCreateModuleImports(Edge lastEdge, out AvailableModuleImports availableImports)
+            => TryCreateModuleImports(lastEdge.GetFirstEdge().End, lastEdge.End, out availableImports);
+
+        private static bool TryCreateModuleImports(Node rootNode, Node moduleNode, out AvailableModuleImports availableImports) {
+            if (moduleNode.IsModule) {
+                availableImports = new AvailableModuleImports(moduleNode.Name, moduleNode.FullModuleName, rootNode.Name, moduleNode.ModulePath);
+                return true;
+            }
+
+            var initPyNode = moduleNode.GetChild("__init__");
+            if (initPyNode != default && initPyNode.IsModule) {
+                availableImports = new AvailableModuleImports(moduleNode.Name, initPyNode.FullModuleName, rootNode.Name, initPyNode.ModulePath);
+                return true;
+            }
+
+            availableImports = default;
+            return false;
+        }
+
+        private static AvailablePackageImports CreateNamespacePackageImports(in Edge lastEdge) {
+            var modules = new List<AvailableModuleImports>();
+            var packageNames = new List<string>();
+            var packageNode = lastEdge.End;
+            var rootNode = lastEdge.GetFirstEdge().End;
+            for (var i = 0; i < packageNode.ChildrenCount; i++) {
+                var child = packageNode.GetChildAt(i);
+                if (TryCreateModuleImports(rootNode, child, out var moduleImports)) {
+                    modules.Add(moduleImports);
+                } else {
+                    packageNames.Add(child.Name);
+                }
+            }
+
+            return new AvailablePackageImports(packageNode.Name, rootNode.Name, modules.ToArray(), packageNames.ToArray());
+        }
+
+        public PathResolverSnapshot SetRoot(in string rootDirectory) {
+            var normalizedRootDirectory = !string.IsNullOrEmpty(rootDirectory) && Path.IsPathRooted(rootDirectory)
+                ? PathUtils.NormalizePath(rootDirectory)
+                : string.Empty;
+            if (_rootDirectory.Equals(normalizedRootDirectory, PathsStringComparison)) {
+                return this;
+            }
+
+            var newRoots = CreateRoots(normalizedRootDirectory, _userSearchPaths, _interpreterSearchPaths);
+            return new PathResolverSnapshot(_pythonLanguageVersion, normalizedRootDirectory, _userSearchPaths, _interpreterSearchPaths, newRoots, Version + 1);
         }
 
-        public PathResolverSnapshot SetSearchPaths(in IEnumerable<string> searchPaths) {
-            var rootNodes = searchPaths.Where(Path.IsPathRooted).Select(Node.CreateRoot).Prepend(_roots[0]).ToArray();
-            return new PathResolverSnapshot(_pythonLanguageVersion, rootNodes, Version + 1);
+        public PathResolverSnapshot SetUserSearchPaths(in IEnumerable<string> searchPaths) {
+            var userSearchPaths = searchPaths.ToArray();
```

Comment:
User search paths can contain anything, like `.` for example

Reply:
PathResolver never accesses file system, so there can't be any IOExceptions. Search paths are preserved in their form cause when `_rootDirectory` is changed, we can try to convert relative path to absolute.

Reply:
I think `Path.Combine(rootDirectory, p)` etc also throw when path parts are invalid, but that is `ArgumentException`